client/trino-jdbc/src/test/java/io/trino/jdbc/TestTrinoResultSet.java

```diff
@@ -0,0 +1,223 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.jdbc;
+
+import com.google.common.collect.ImmutableList;
+import io.trino.client.ClientSelectedRole;
+import io.trino.client.QueryData;
+import io.trino.client.QueryStatusInfo;
+import io.trino.client.StatementClient;
+import io.trino.client.StatementStats;
+import org.testng.annotations.Test;
+
+import javax.annotation.Nullable;
+
+import java.time.ZoneId;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+import static java.util.concurrent.TimeUnit.MILLISECONDS;
+import static org.testng.Assert.assertTrue;
+
+public class TestTrinoResultSet
+{
+    @Test
+    public void testIteratorCancel()
+            throws Exception
+    {
+        MockAsyncIterator<Iterable<List<Object>>> iterator = new MockAsyncIterator<>(new Iterator<Iterable<List<Object>>>()
+        {
+            @Override
+            public boolean hasNext()
+            {
+                return true;
+            }
+
+            @Override
+            public Iterable<List<Object>> next()
+            {
+                return ImmutableList.of((ImmutableList.of(new Object())));
+            }
+        });
+
+        MILLISECONDS.sleep(100);
```

Comment:
Tests with waits in them make me cringe.
A slow machine/VM can make it fail.

Reply:
Indeed. The test should get the code to the expected synchronization points by means of CountdownLatch/Semaphore/CyclicBarrier/etc.

Reply:
Updated. A CountDownLatch is used to notify the main thread instead of the indeterminate Thread.sleep().